### PR TITLE
backend-plugin-api: rework service factory definitions

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

Work towards #13082, see the changeset for details. Main purpose of this is to make service factory implementations cleaner and less confusing, as well as support TypeScript 4.9.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by freben** [comment]
It seems to claim there's too many changesets :)

File: packages/backend-plugin-api/api-report.md [hunk 1]
```diff
@@ -229,6 +293,30 @@ export interface PluginMetadataService {
...
+  TContext,
+  TImpl extends TService,
+  TDeps extends {
+    [name in string]: ServiceRef<unknown>;
+  },
+> {
+  // (undocumented)
+  createRootContext?(
+    deps: ServiceRefsToInstances<TDeps, 'root'>,
+  ): Promise<TContext>;
```
Comment (freben) [inline packages/backend-plugin-api/api-report.md:308]: Honestly now that we've simplified the structure of this interface (as in, the formerly more nested return types), I think we should at the same time make them return `T | Promise<T>`. Same for the factory itself. I imagine especially for the context maker, that it'll almost never _want_ to be async. It'll be a nice convenience, at little-to-no complexity cost for the framework

File: packages/backend-plugin-api/api-report.md [hunk 2]
```diff
@@ -126,22 +126,86 @@ export function createExtensionPoint<T>(
...
+  TDeps extends {
+    [name in string]: ServiceRef<unknown>;
+  },
+  TOpts extends object | undefined = undefined,
+>(
+  config: RootServiceFactoryConfig<TService, TImpl, TDeps>,
+): () => ServiceFactory<TService>;
+
+// @public
+export function createServiceFactory<
```
Comment (freben) [inline packages/backend-plugin-api/api-report.md:142]: So many duplicates! But yeah that also should make things easier for the compiler

File: packages/backend-plugin-api/api-report.md (hunk 1 above)
Comment (benjdlambert) [reply packages/backend-plugin-api/api-report.md:308]: Yeah I saw this pattern when refactoring and thought the same. Maybe it's worth supporting.

File: packages/backend-plugin-api/api-report.md (hunk 2 above)
Comment (Rugvip) [reply packages/backend-plugin-api/api-report.md:142]: Yeah they're all needed to make things work 😅

File: packages/backend-app-api/src/services/implementations/tokenManager/tokenManagerFactory.ts [hunk 3]
```diff
@@ -31,6 +31,6 @@ export const tokenManagerFactory = createServiceFactory({
     const tokenManager = ServerTokenManager.fromConfig(config, {
       logger,
     });
-    return async () => tokenManager;
+    return tokenManager;
```
Comment (freben) [inline packages/backend-app-api/src/services/implementations/tokenManager/tokenManagerFactory.ts:34]: Just noting that this is a change of behavior; it'll make multiple instances now but didn't before. Maybe that's safe here

**Review by freben — APPROVED** [review]
Amazeballs

File: packages/backend-app-api/src/services/implementations/tokenManager/tokenManagerFactory.ts (hunk 3 above)
Comment (benjdlambert) [reply packages/backend-app-api/src/services/implementations/tokenManager/tokenManagerFactory.ts:34]: Hmm does this need moving to `createRootContext` then I wonder? Didn't we have a similar problem last week with duplicated `TokenManagers`?

Comment (Rugvip) [reply packages/backend-app-api/src/services/implementations/tokenManager/tokenManagerFactory.ts:36]: 👍 , fixed + added a regression test

**Review by benjdlambert — APPROVED** [review]

**Pull request merged**